Update docs about adding template (#1674)
CompuIves/codesandbox-client: contributor-docs/adding-template.md

# Contribute with a Template

> To contribute with a CodSandbox template, you need to complete multiple steps and submit two Pull Requests and we have created this guide to help you do this. Here you will find descriptions and explanations for what you need to do, along with some examples that we have added as references.

Reviewer: ```suggestion
To contribute with a CodeSandbox template, you need to complete multiple steps and submit two Pull Requests and we have created this guide to help you do this. Here you will find descriptions and explanations for what you need to do, along with some examples that we have added as references.
```

> To contribute with a CodSandbox template, you need to complete multiple steps and submit two Pull Requests and we have created this guide to help you do thisTo do this. Here you will find descriptions and explainations for what you need to do, along with some examples that we have added as references.

Reviewer 2: ```suggestion
To contribute with a CodSandbox template, you need to complete multiple steps and submit two Pull Requests and we have created this guide to help you do this. Here you will find descriptions and explanations for what you need to do, along with some examples that we have added as references.
```

We realise that the process of submitting a template is not straightforward and we hope that this guide will help you along the way. We are working on a system that will make it easier.

If you think that we missed something in this guide or believe we could explain something better, please let us know by submitting an [issue](https://github.com/CompuIves/codesandbox-client/issues/new/choose) with your feedback.

## What is a template?

> A template is an identifier for a specific type of sandbox project that you can create on [codesandbox.io](https://codesandbox.io), like projects using `Gatsby`, `React` or `Vue`.

Reviewer: ```suggestion
A template is an identifier for a specific type of sandbox project that you can create on [codesandbox.io](https://codesandbox.io), like projects using `Gatsby`, `React` or `Vue.js`.
```

When you create a template you can customise how the template behaves, in order to improve the User Experience of your template in the CodeSandbox editor and preview. Examples of this include configuring which file should be opened in the editor by default when choosing a template, or changing the default `eslint` rules in a template, like with `vue-cli`.

We encourage template creators to improve the editor experience for their templates in order to give everyone the best possible experience when using templates.

## Template types

Templates can be of different types: **Sandboxes** or **Containers**, and have major differences in functionalities. It's important to know these differences before you start working on a new template.

### Sandboxes

CodeSandbox executes projects in the browser, which we call sandboxes. This means that the transpiling, bundling, dependency resolvment and more happens in the browser itself, without a server being involved. This has some advantages over conventional approaches; it works offline, is more performant and doesn't give us server costs, which means we can have many sandboxes without having to worry (a lot ;-)).

> There are also some disadvantages to this approach. When sandboxes run in the browser, we lose flexibility. It's no longer possible to run Command Line Interface (CLI) commans and in some cases custom configations are not supported, for example a custom webpack configuration. This is why we have developed a new kind of sandbox called **Container**, which we released in September 2018.

Reviewer 3: ```suggestion
There are also some disadvantages to this approach. When sandboxes run in the browser, we lose flexibility. It's no longer possible to run Command Line Interface (CLI) commands and in some cases custom configurations are not supported, for example a custom webpack configuration. This is why we have developed a new kind of sandbox called **Container**, which we released in September 2018.
```

### Containers

Unlike sandboxes, **containers** are executed on a server. This makes it possible to create projects that are end-to-end based, like `Next.js` with CodeSandbox, and also makes it possible to build bigger projects. **Containers** let you run any command, and everything that works locally will also work in container.

> However, like the sandboxes, **containers** also come with some limitations. In order to work on a container, you need to be signed in as a user, you can't edit containers while offline, it's not possible viewing them from an embed, and you can only have a limited amount of container based projects.

Reviewer 3: ```suggestion
However, like the sandboxes, **containers** also come with some limitations. In order to work on a container, you need to be signed in as a user, you can't edit containers while offline, it's not possible to edit them from an embed, and you can only have a limited amount of container based projects.
```

> However, like the sandboxes, **containers** also come with some limitations. In order to work on a container, you need to be signed in as a user, you can't edit containers while offline, it's not possible viewing them from an embed, and ~~you can only have a limited amount of container based projects~~ ([containers are now out of beta](https://medium.com/@compuives/codesandbox-containers-out-of-beta-improvements-be79679df4cc)).

Reviewer 2: Containers are out of beta but we still have a limited amount a user can have :/

Reviewer: What @SaraVieira said, we should drop the strike-through to not mislead people. As a side-note, client sandboxes are also limited for regular users, only the limit is higher (50 vs. 15).

Author: Fixed 😊

### Which one to choose?

As you have read above, the template type determines whether a project is executed in a *sandbox in a browser* or in a *container on a server*. This means that the template type you have to choose, depends on your specific use-case and where you want your project to be executed.

> If you would like to demonstrate CLI functionality, we recommend that you use a **container** type template, and if you want to demonstrate a JavaScript framework (like `create-react-app`) we recommended using a**sandbox** type template.

Reviewer 2: ```suggestion
If you would like to demonstrate CLI functionality, we recommend that you use a **container** type template, and if you want to demonstrate a JavaScript framework (like `react`) we recommended using a**sandbox** type template.
```

Create React App is the CLI, maybe makes sense for this to be react

Reviewer: `React`, capital "r". 🙂

We encourage everyone to first evaluate whether the template works as a sandbox, before deciding on using a container.

## Adding a new template

In order to add a new template, you need to go through a set of steps. Some of these steps you have to do, others depend on the type of template you want to add (**sandbox** vs. **container**).

> To get started, you should first follow the steps in our Contribution guidelines in order to [set up CodeSandbox locally](https://github.com/CompuIves/codesandbox-client/blob/master/CONTRIBUTING.md#setting-up-the-project-locally).

Reviewer: ```suggestion
To get started, you should first follow the steps in our contribution guidelines in order to [set up CodeSandbox locally](https://github.com/CompuIves/codesandbox-client/blob/master/CONTRIBUTING.md#setting-up-the-project-locally).
```

### 1. Add template logo

Add the logo for your template in the [logos directory](https://github.com/CompuIves/codesandbox-client/tree/master/packages/common/src/components/logos) (`codesandbox-client/packages/common/src/components/logos`). How you do this depends on the file format of your logo.

#### SVG logos

Create a `.tsx` file in the `/logos` directory with the appropirate name and content. If your template's name is "Banana", name your logo file "Banana".

Examples:

- [Vue logo](https://github.com/CompuIves/codesandbox-client/blob/master/packages/common/src/components/logos/Vue.tsx)
- [React logo](https://github.com/CompuIves/codesandbox-client/blob/master/packages/common/src/components/logos/React.tsx)

#### PNG/JPEG logos

Add the image file in the `/logos` directory and create `.tsx` file with the appropriate name.

Example:

```js
import * as React from 'react';
import filename from './filename.png';

export default props => <img alt="filename" src={filename} {...props} />;
```

> Replace `filename` with your logo name. The file name should be all lowercase and the `.tsx` file should be capitalised: `Filename.tsx`. 

Reviewer: ```suggestion
Replace `filename` with your logo name. The image file name should be all lowercase and the `.tsx` file should be capitalised: `Filename.tsx`. 
```
